### PR TITLE
Fix cancelable_action_client example

### PR DESCRIPTION
**Pull request by msimberg**
Fixes errors like this:

```
terminating with uncaught exception of type hpx::thread_interrupted: std::exception
```

in the `cancelable_action_client` example.

## Proposed Changes

- Use atomic to protect `id_` instead of spinlock mutex

## Any background context you want to provide?

The commit message contains more details on what happens.

Side note: this is triggered because of `spinlock` calling `suspend` after a few iterations (in `yield_k`). `spinlock` to me sounds like it should never suspend the current thread. Should that be changed as well?

### CODE REVIEW

**Comment by sithhell** [comment]
Regarding the call to suspend: This is actually a misnomer ... what it really does is yielding the current thread. no suspension going on there...

File: examples/cancelable_action/cancelable_action/server/cancelable_action.hpp [hunk 1]
```diff
@@ -67,19 +70,18 @@ namespace examples { namespace server
         // Cancel the lengthy action above
         void cancel_it()
         {
-            std::lock_guard<mutex_type> l(mtx_);
-            if (id_ != hpx::thread::id()) {
-                hpx::thread::interrupt(id_);
```
Comment (sithhell) [inline examples/cancelable_action/cancelable_action/server/cancelable_action.hpp]: hmm, wouldn't it make more sense to unlock `mtx_` before calling `hpx::thread::interrupt(id_)`?
Something like this:
```

hpx::thread_id::id id;
{
    std::lock_guard<mutex_type> l(mtx_);
    std::swap(id, id_);
}
if (id != hpx::thread::id())
{
    hpx::thread::interrupt(id);
}
````

Comment (msimberg) [reply examples/cancelable_action/cancelable_action/server/cancelable_action.hpp]: Yeah, that could be changed, but it's actually taking the lock in the `reset_id` destructor that causes the problem. To be honest, the example wouldn't even need an atomic (in this particular example and use case).

**Comment by msimberg** [comment]
> Regarding the call to suspend: This is actually a misnomer ... what it really does is yielding the current thread. no suspension going on there...

True. However, the point is that `suspend` and `yield` both call `interruption_point`. And `yield_k` also eventually does timed suspension.

**Comment by sithhell** [comment]
> > Regarding the call to suspend: This is actually a misnomer ... what it really does is yielding the current thread. no suspension going on there...
> 
> True. However, the point is that `suspend` and `yield` both call `interruption_point`. And `yield_k` also eventually does timed suspension.

fair enough.

File: examples/cancelable_action/cancelable_action/server/cancelable_action.hpp (hunk 1 above)
Comment (sithhell) [reply examples/cancelable_action/cancelable_action/server/cancelable_action.hpp]: Ok, I haven't looked too closely. Just noticed the slight change of semantics with your proposed changes...

**Pull request merged**